src/nfa/nfa-intersection.cc

```diff
@@ -119,27 +120,51 @@ class Intersection
     /**
      * Compute intersection preserving epsilon transitions.
      */
-    void compute_preserving_epsilon_transitions()
-    {
+    void compute_preserving_epsilon_transitions() {
         initialize_pairs_to_process();
 
-        if (lhs.trans_empty() || rhs.trans_empty()) { return; }
-
-        while (!pairs_to_process.empty())
-        {
+        while (!pairs_to_process.empty()) {
             pair_to_process = *pairs_to_process.begin();
             pairs_to_process.erase(pair_to_process);
+            compute_transitions_for_eps_pres_prod_for_state_pair();
+        }
+    }
 
-            // find all transitions that have same symbol for first and the second state in the pair_to_process
-            for (const auto& this_state_transitions: lhs.transitionrelation[pair_to_process.first] )
-            {
-                for (const auto& other_state_transitions: rhs.transitionrelation[pair_to_process.second])
-                {
-                    compute_for_state_pair_preserving_eps_trans(this_state_transitions,
-                                                                other_state_transitions);
+    /**
+     * Compute transitions for epsilon preserving product for a current state pair.
+     */
+    void compute_transitions_for_eps_pres_prod_for_state_pair() {
+        for (const auto& lhs_state_transitions: lhs.transitionrelation[pair_to_process.first]) {
+            // Check for lhs epsilon transitions.
+            if (lhs_state_transitions.symbol == epsilon) {
+                compute_for_lhs_state_epsilon_transitions(lhs_state_transitions);
```

Comment:
I'm not fully familiar with the algorithm, however, I feel, that there could be `break` (or not; just please check)

Reply:
Why would there be `break`? This is an algorithm for intersection preserving epsilon transitions. I want the epsilon transitions to be added. Here, I add lhs epsilon transitions first, but I want to possibly add rhs epsilon transitions in case the rhs state also has epsilon transitions and/or both states (lhs state and rhs state) have epsilon transitions. Therefore, I continue after adding the lhs transitions with adding the “normal transitions” and rhs epsilon transitions.

Reply:
Sorry, I actually meant to say `continue` instead of `break` :). I wasn't even suggesting this is error, just that I felt, it might be the case, that if `symbol == epsilon`, maybe the rest of the for loop would not neet to be executed.

Reply:
That would make more sense. But it is not the intended behaviour here (as explained above). I am closing this discussion as resolved.

Reply:
Yes, that was my point: whether this is intenteded or not.  I'm closing this then.